pkg/handlers/defaulthandler.go

```diff
@@ -0,0 +1,235 @@
+// This function is used to create an Option that sets the skipArchives field of a DefaultHandler.
+// If skip is true, the DefaultHandler will skip archive files.
+func WithSkipArchives(skip bool) Option {
+	return func(h *DefaultHandler) { h.skipArchives = skip }
+
+}
+
+// DefaultHandler provides a base implementation for file handlers, encapsulating common behaviors
+// needed across different handlers, such as skipping binary files. This handler is embedded in other
+// specialized handlers to ensure consistent application of these common behaviors and to simplify
+// the extension of handler functionalities.
+type DefaultHandler struct {
+	skipBinaries bool // Controls whether binary files should be processed.
+	skipArchives bool // Controls whether archive files should be processed.
+}
+
+// configure applies a series of Options to the handler, allowing for flexible configuration of
+// handler-specific settings such as whether to skip binary files. This method supports dynamic
+// customization of handler behavior at runtime.
+func (h *DefaultHandler) configure(opts ...Option) {
+	for _, opt := range opts {
+		opt(h)
+	}
+}
+
+// HandleFile initiates the asynchronous extraction of archive contents. It creates a channel to send extracted data
+// and starts a goroutine that handles the archive opening process. It utilizes a context with a timeout to ensure
+// that the extraction process does not exceed a predefined maximum duration.
+// The function returns a channel that will receive the extracted data bytes and an error if the initial setup fails.
+func (h *DefaultHandler) HandleFile(ctx logContext.Context, input *diskbufferreader.DiskBufferReader) (chan []byte, error) {
+	if h.skipArchives {
+		return nil, nil
+	}
+
+	archiveChan := make(chan []byte, defaultBufferSize)
+	go func() {
+		ctx, cancel := logContext.WithTimeout(ctx, maxTimeout)
+		defer cancel()
+		defer close(archiveChan)
+
+		if err := h.openArchive(ctx, 0, input, archiveChan); err != nil {
+			ctx.Logger().Error(err, "error unarchiving chunk.")
+		}
+	}()
+	return archiveChan, nil
+}
+
+var ErrMaxDepthReached = errors.New("max archive depth reached")
+
+// openArchive recursively extracts content from an archive up to a maximum depth, handling nested archives if necessary.
+// It takes a reader from which it attempts to identify and process the archive format. Depending on the archive type,
+// it either decompresses or extracts the contents directly, sending data to the provided channel.
+// Returns an error if the archive cannot be processed due to issues like exceeding maximum depth or unsupported formats.
+func (h *DefaultHandler) openArchive(ctx logContext.Context, depth int, reader io.Reader, archiveChan chan []byte) error {
+	if common.IsDone(ctx) {
+		return ctx.Err()
+	}
+
+	if depth >= maxDepth {
+		return ErrMaxDepthReached
+	}
+
+	format, arReader, err := archiver.Identify("", reader)
+	if errors.Is(err, archiver.ErrNoMatch) && depth > 0 {
```

Comment:
What is this depth check for?

Reply:
I believe this configuration controls the maximum depth of nested archives that we support. It gets set [here](https://github.com/trufflesecurity/trufflehog/blob/4a5fbf84178018da2d92aa4d0a236add497da27b/main.go#L66).

I’m fairly confident that I haven’t altered the existing functionality. However, let me know if there are any unintended changes.

Reply:
Oh, this occurs because when the depth is greater than 0 and the `archiver.Identify` function fails to identify a format, it indicates that the `openArchive` function was invoked recursively. Under these conditions, the new reader has been successfully extracted/decompressed.

Reply:
At the very least I think this deserves a comment, because I think it's very non-obvious. But I also think it's worth taking a step back and seeing if there's a more straightforward way to write this. In its current state, `openArchive` expects a reader that represents either archived content or non-archived content that's been retrieved from inside an archive. That's a pretty idiosyncratic precondition, and I find it difficult to reason about. At the very least, would it be possible to lift this check up into the caller so that `openArchive` treats all passed-in readers the same, irrespective of the depth they were extracted at?

Reply:
I agree with your points 100%. My hesitation stems from trying to avoid unnecessary changes in this refactor PR. I’m willing to add a comment to clarify this aspect. However, I suggest handling the check relocation in a separate PR to keep the current one focused and free from new logic. Thoughts?